### PR TITLE
Handle Statistics and SparseArrays as extensions

### PR DESCRIPTION
**Pull request by dkarrasch**
I went ahead and just did it. The only "disturbing" thing is that `kron` method which is not an extension, strictly speaking. So it would be great to resolve it and make the dependencies weak.

### CODE REVIEW

File: ext/FillArraysSparseArraysExt.jl [hunk 1]
```diff
@@ -0,0 +1,57 @@
+module FillArraysSparseArraysExt
+
+using SparseArrays
```
Comment (jishnub) [inline ext/FillArraysSparseArraysExt.jl:3]: You'll probably need
```diff
+ using LinearAlgebra
```
and
```diff
- using FillArrays: RectDiagonalFill, RectOrDiagonalFill, ZerosVector, ZerosMatrix
+ using FillArrays: RectDiagonalFill, RectOrDiagonalFill, ZerosVector, ZerosMatrix, getindex_value
```

**Comment by jishnub** [comment]
The Aqua formatting error on v1.6 should be fixed by https://github.com/JuliaArrays/FillArrays.jl/pull/287

**Comment by jishnub** [comment]
Locally, I can get around the `kron` issue using
`fillalgebra.jl`:
```julia
abstract type KronStyle end
struct SparseKronStyle <: KronStyle end
struct UnknownKronStyle <: KronStyle end
KronStyle(A) = UnknownKronStyle()
kron(A::RectDiagonalFill, B::RectDiagonalFill) =
    maybesparsekron(KronStyle(A), KronStyle(B), A, B)

function maybesparsekron(::UnknownKronStyle, ::UnknownKronStyle, A::AbstractMatrix, B::AbstractMatrix)
    invoke(kron, Tuple{AbstractMatrix, AbstractMatrix}, A, B)
end
```
`FillArraysSparseArraysExt.jl`:
```julia
FillArrays.KronStyle(A::RectDiagonalFill) = FillArrays.SparseKronStyle()

function FillArrays.maybesparsekron(::SparseKronStyle, ::SparseKronStyle, A, B)
    kron(sparse(A), sparse(B))
end
```

But, this is a more serious variant of type-piracy, where behavior is altered based on whether a library is loaded (and is potentially also a regression, as currently working code may lead to OOM errors if one doesn't load `SparseArrays`). Nonetheless, this may not be a widely used code path, so maybe such a change is acceptable?

**Comment by dlfivefifty** [comment]
The `kron` overload is a bit funny. It might be better to make a custom type if this is actually needed. Otherwise deprecate it and users can write `kron(sparse(A), sparse(B))` instead?

**Comment by dkarrasch** [comment]
In fact, one should really go with a lazy Kronecker product, as in `LazyArrays.jl`, or `Kronecker.jl`, or `LinearMaps.jl` if you don't need the result to subtype `AbstractMatrix`. With the "vec trick" and fast multiplication, this should be superior in most applications.

**Comment by dkarrasch** [comment]
How do we proceed here? Did you mean to release one version with a deprecation only, and then a subsequent version with the Pkg extension stuff like here? An alternative would be to release v2, which may take a while to propagate through the ecosystem.

**Comment by dlfivefifty** [comment]
Technically for SEMVAR: if we wanted to drop support we could deprecate in a `1.x` release and then remove in a `2.0` release. But then this means there's no actual benefit until 2.0!

BUT: I take the view that a change is non-breaking if it doesn't break anything. @jishnub why was `kron` added? Do you need it or could you use `kron(sparse(A), sparse(B))`?

**Comment by jishnub** [comment]
`kron` was added because there was an issue about this: https://github.com/JuliaArrays/FillArrays.jl/issues/265. I don't have strong opinions either way.

**Comment by dlfivefifty** [comment]
How about this: just have `kron` throw an error when SparseArrays.jl is not loaded? I.e.
```julia
# in FillArrays.jl
kron(A::RectDiagonalFill, B::RectDiagonalFill) = kron_fill(A, B)
kron_fill(A, B) = error("Load SparseArrays.jl")

# in FillArraysSparseArraysExt.jl
kron_fill(A::RectDiagonalFill, B::RectDiagonalFill) = kron(sparse(A), sparse(B))
```

**Comment by jishnub** [comment]
This would be breaking, though. The hack-y solution that I had presented above was trying to work around this breakage.

**Comment by dlfivefifty** [comment]
It's not really "breaking" as downstream one just has to do `using SparseArrays`. I'd be shocked if there's any extant code that's impacted

**Comment by jishnub** [comment]
Alright, perhaps we may go ahead with making this error without `SparseArrays`

**Comment by jishnub** [comment]
I think this should be good to merge?

**Comment by amilsted** [comment]
Do I understand right that `kron` will now be deprecated for these arrays regardless of whether `SparseArrays` is loaded? That seems fair enough. It's a bit weird that the extension package changes the output type of `kron`, but if it's going away I suppose it's not a big deal.

**Comment by jishnub** [comment]
Yes, that's the status quo

**Pull request merged**